src/models/survey/question.cr

```diff
@@ -12,13 +12,19 @@ class Survey
     column choices : JSON::Any, presence: false
     column max_rating : Int32?
     column tags : Array(String), presence: false
+    column deleted_at : Int64?
 
     has_many answers : Survey::Answer, foreign_key: "answer_id"
 
     timestamps
 
-    # TODO: check question is not used by a survey before deleting
-    # before :delete, :ensure_not_used
+    before(:save) do |m|
+      question_model = m.as(Question)
+      if question_model.persisted? && Survey::Answer.query.where(question_id: question_model.id).count > 0
+        question_model.soft_delete
+        question_model.clear_persisted
```

Comment:
`question_model.persisted?` is `true` if the model is in the database.

`Survey::Answer.query.where(question_id: question_model.id).count > 0` is `true` if there are any linked answers.

So if the model is in the database and there are linked answers, then:

- `question_model.soft_delete` we mark it as deleted
- `question_model.clear_persisted` clears the persistent state so that our changed model gets inserted as a new row in the database and gets a new id.

By doing this we ensure that questions with linked answers are never changed as changing the question could change the meaning of the answers.

Reply:
Basically, Clear does not have the ability to cancel update/insert in the lifecycle hooks, so you either have to work around that or move the logic to the controllers (putting it in the controllers, feels like a future bug waiting to happen).

Reply:
`question_model.soft_delete` uses `Clear::SQL.execute Clear::SQL.raw(...)` to bypass the lifecycle hooks. If it did not bypass the lifecycle hooks then it would create an infinite loop.